lib/rtp_vp9/vp9_payload_descriptor.ex

```diff
@@ -219,64 +198,62 @@ defmodule Membrane.RTP.VP9.PayloadDescriptor do
        when diff_count < 3 do
     <<_::7, n::1>> = p_diff
 
-    if n == 1 do
-      case get_pdiffs(header, rest, diff_count + 1, descriptor_acc) do
-        {:ok, {descriptor_acc, rest}} ->
-          {:ok,
-           {%{
-              descriptor_acc
-              | p_diffs: [:binary.decode_unsigned(p_diff) | descriptor_acc.p_diffs]
-            }, rest}}
-
-        _ ->
-          {:error, :malformed_data}
-      end
-    else
+    with 1 <- n,
+         {:ok, {descriptor_acc, rest}} <-
+           get_pdiffs(header, rest, diff_count + 1, descriptor_acc) do
       {:ok,
-       {%{descriptor_acc | p_diffs: [:binary.decode_unsigned(p_diff) | descriptor_acc.p_diffs]},
-        rest}}
+       {%{
+          descriptor_acc
+          | p_diffs: [:binary.decode_unsigned(p_diff) | descriptor_acc.p_diffs]
+        }, rest}}
+    else
+      0 ->
+        {:ok,
+         {%{descriptor_acc | p_diffs: [:binary.decode_unsigned(p_diff) | descriptor_acc.p_diffs]},
+          rest}}
+
+      {:error, :malformed_data} ->
+        {:error, :malformed_data}
     end
   end
 
-  defp get_pdiffs(_, rest, _, descriptor_acc) when byte_size(rest) > 0,
+  defp get_pdiffs(_header, rest, _diff_count, descriptor_acc) when byte_size(rest) > 0,
     do: {:ok, {descriptor_acc, rest}}
```

Comment:
why does that return what it gets?

Reply:
According to documentation P_DIFFS are present only when both P and F are set to one and in such case `header` will match `<<_i::1, 1::1, _l::1, 1::1, _::4>>`. Otherwise parser doesn't cut P_DIFFS form binary so it is returned unchanged.

[https://tools.ietf.org/html/draft-ietf-payload-vp9-10#section-4.2](url)
![image](https://user-images.githubusercontent.com/27630111/102909039-a0d4d880-4478-11eb-9485-79724bb0f55f.png)